src/DiagonalHessianApproximation.jl

```diff
@@ -0,0 +1,222 @@
+  nrow::I
+  ncol::I
+  symmetric::Bool
+  hermitian::Bool
+  prod!
+  tprod!
+  ctprod!
+  nprod::I
+  ntprod::I
+  nctprod::I
+  args5::Bool
+  use_prod5!::Bool # true for 5-args mul! and for composite operators created with operators that use the 3-args mul!
+  Mv5::V
+  Mtu5::V
+  allocated5::Bool # true for 5-args mul!, false for 3-args mul! until the vectors are allocated
+end
+
+SpectralGradient(d::T, n::I, V) where {T <: Real, I <: Integer} = 
+  SpectralGradient(
+    d,
+    n,
+    n,
+    true, 
+    true,  
+    (res, v, α, β) -> mulSquareOpDiagonal!(res, d, v, α, β), 
+    (res, v, α, β) -> mulSquareOpDiagonal!(res, d, v, α, β), 
+    (res, v, α, β) -> mulSquareOpDiagonal!(res, d, v, α, β), 
+    0,
+    0,
+    0,
+    true,
+    true,
+    V(undef,0),
+    V(undef,0),
+    true)
+
+# update function
+# s = x_{k+1} - x_k
+# y = ∇f(x_{k+1}) - ∇f(x_k)
+function push!(
+  B::SpectralGradient{T,I,V},
+  s::V,
+  y::V
+  ) where {T <: Real, I <: Integer, V <: AbstractVector{T}}
+  if s .== 0
+    error("Cannot divide by zero and s .= 0")
+  end
+  B.d = dot(s,y)/dot(s,s)
+end
+
+"""
+Implementation of a modified SR1 method described in
+
+Farzin Modarres, Abu Hassan Malik, Wah June Leong,
+Improved Hessian approximation with modified secant equations for symmetric rank-one method.
+https://doi.org/10.1016/j.cam.2010.10.042.
+"""
+
+mutable struct DiagonalModifiedSR1{T <: Real, I <: Integer, V <: AbstractVector{T}} <: AbstractDiagonalQuasiNewtonOperator{T} 
+  d::V # Diagonal of the operator
+  yt::V # Preallocate a vector used in push! function
+  Bs::V # Preallocate a vector used in push! function
+  nrow::I
+  ncol::I
+  symmetric::Bool
+  hermitian::Bool
+  prod!
+  tprod!
+  ctprod!
+  nprod::I
+  ntprod::I
+  nctprod::I
+  args5::Bool
+  use_prod5!::Bool # true for 5-args mul! and for composite operators created with operators that use the 3-args mul!
+  Mv5::V
+  Mtu5::V
+  allocated5::Bool # true for 5-args mul!, false for 3-args mul! until the vectors are allocated
+end
+
+DiagonalModifiedSR1(d::AbstractVector{T}) where {T <: Real} = 
+  DiagonalModifiedSR1(
+    d,
+    d,
+    d,
+    length(d),
+    length(d),
+    true, 
+    true,  
+    (res, v, α, β) -> mulSquareOpDiagonal!(res, d, v, α, β), 
+    (res, v, α, β) -> mulSquareOpDiagonal!(res, d, v, α, β), 
+    (res, v, α, β) -> mulSquareOpDiagonal!(res, d, v, α, β), 
+    0,
+    0,
+    0,
+    true,
+    true,
+    typeof(d)(undef,0),
+    typeof(d)(undef,0),
+    true)
+
+# update function
+# s = x_{k+1} - x_k
+# y = ∇f(x_{k+1}) - ∇f(x_k)
+# t = ∇f(x_k) + ∇f(x_{k+1})
+# z = f(x_k) - f(x_{k+1})
+# u ∈ {s,y,∇f(x_k)}
+function push!(
+  B::DiagonalModifiedSR1{T,I,V},
+  s::V,
+  y::V,
+  t::V,
+  z::T,
+  u::V
+  ) where {T <: Real, I <: Integer, V <: AbstractVector{T}}
+  ψ = 2 * z + dot(t,s)
+  B.Bs .= B*s
+  sTu = dot(s,u)
+  if sTu == 0
+    error("Cannot divide by zero and dot(s,u) = 0")
+  end
+  B.yt .= y .+ abs(ψ)/dot(s,u) .* u .- B.Bs
```

Comment:
`dot(s, u)` already computed
```suggestion
  B.yt .= y .+ abs(ψ)/sTu .* u .- B.Bs
```